scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/bsp/server/connection/ScalaConnectionDetailsProvider.scala

```diff
@@ -0,0 +1,127 @@
+import org.jetbrains.bsp.protocol.utils.ParsersKt
+import org.jetbrains.plugins.bsp.extension.points.BuildToolId
+import org.jetbrains.plugins.bsp.server.connection.ConnectionDetailsProviderExtension
+import org.jetbrains.plugins.scala.bsp.config.ScalaPluginConstants.BUILD_TOOL_ID
+
+import java.lang
+import java.util.concurrent.TimeUnit
+import scala.jdk.CollectionConverters._
+
+class ScalaConnectionDetailsProvider extends ConnectionDetailsProviderExtension {
+  override def onFirstOpening(project: Project, projectPath: VirtualFile, continuation: Continuation[_ >: lang.Boolean]): AnyRef = {
+    //noinspection ApiStatus,UnstableApiUsage
+    JavaCoroutines.suspendJava[lang.Boolean](cont => {
+      saveProjectRootFile(project, projectPath)
+      val connectionFile = getConnectionFile(projectPath)
+      if (connectionFile.isDefined) {
+        cont.resume(true)
+      }
+
+      val hasGeneratedFile = generateSbtConnectionFile(projectPath).isDefined
+      cont.resume(hasGeneratedFile)
+    }, continuation)
+  }
+
+  override def provideNewConnectionDetails(project: Project, bspConnectionDetails: BspConnectionDetails): BspConnectionDetails = {
+    if (bspConnectionDetails != null) {
+      return bspConnectionDetails
+    }
+
+    val projectRootPath = getProjectRootFile(project) match {
+      case Some(projectFile) => projectFile
+      case None =>
+        error("Cannot obtain project path, please reimport the project.")
+        return bspConnectionDetails
+    }
+
+    val connectionFile = getConnectionFile(projectRootPath) match {
+      case None => generateSbtConnectionFile(projectRootPath)
+      case other => other
+    }
+
+    connectionFile match {
+      case None => null
+      case Some(connectionFile) => parseConnectionFile(connectionFile).orNull
+    }
+  }
+
+
+  override def getBuildToolId: BuildToolId = BUILD_TOOL_ID
+
+  private def parseConnectionFile(connectionFile: VirtualFile): Option[BspConnectionDetails] =
+    Option(ParsersKt.parseBspConnectionDetails(connectionFile))
+
+  private def saveProjectRootFile(project: Project, projectRootFile: VirtualFile): Unit =
+    project.getActualComponentManager.getService(classOf[ScalaConnectionDetailsProviderService])
+      .loadState(new ScalaConnectionDetailsProviderMetadata(Option(projectRootFile)))
+
+  private def getProjectRootFile(project: Project): Option[VirtualFile] = {
+    project.getActualComponentManager.getService(classOf[ScalaConnectionDetailsProviderService])
+      .getState
+      .projectRootPath
+  }
+
+  private def getConnectionFile(projectPath: VirtualFile): Option[VirtualFile] = {
+    getChild(projectPath, List.apply(".bsp", "sbt.json"))
+  }
+
+  private def generateSbtConnectionFile(projectPath: VirtualFile): Option[VirtualFile] = {
+    val process = new ProcessBuilder("coursier", "launch", "sbt", "--", "bspConfig")
+      .directory(projectPath.toNioPath.toFile)
+      .start
+
+    if (process.waitFor(2, TimeUnit.MINUTES)) {
```

Comment:
hmm, it's going to block the thread

Reply:
Java's process is blocking by default, and we will have to block it in order to generate connection file. What are your concerns about that? Will it break some code from the intellij-bsp?